Solved bug with ClassComparatar relating to Method / Constructor Overloading

Caleb-Wishart opened the pull request:
It appears that in situations where method / Constructor overloading occurs with only a single variable type, chalkbox will get confused with conformance.

The following tests were completed using chalkbox-0.5.3 and the following base class (placed in the packaged mentioned).

```java
package tms.intersection;

public class Coordinate {

    // default
    public Coordinate(){}
    // single int
    public Coordinate(int x){}
    // double int
    public Coordinate(int x, int y){}
    // default
    public void test(){}
    // single int
    public void test(int x){}
    // double int
    public void test(int x, int y){}
}

```
With the base class the following results were given using the JavaEngine testEngineBasic

```
tms.intersection.Coordinate does not conform:
class tms.intersection.Coordinate
    Method tms.intersection.Coordinate.test
        Method parameters do not match!
        Missing: [int, int]
        Extra:   []

    Constructor tms.intersection.Coordinate.tms.intersection.Coordinate
        Constructor parameters do not match!
        Missing: [int, int]
        Extra:   []
```

Some tests were done to try and determine the cause. (In all tests both constructors and method are affected due to the similar nature of how they are checked. A diff is left to the exercise of the reader).

1. Remove any of the 3 constructors
```
tms.intersection.Coordinate conforms
```
2. Remove single int but add triple int
```
tms.intersection.Coordinate does not conform:
class tms.intersection.Coordinate
    Constructor tms.intersection.Coordinate.tms.intersection.Coordinate
        Constructor parameters do not match!
        Missing: [int, int, int]
        Extra:   []
```
3. Change single int to single String
```
tms.intersection.Coordinate conforms
```
4. Change all ints to String
```
tms.intersection.Coordinate does not conform:
class tms.intersection.Coordinate
    Constructor tms.intersection.Coordinate.tms.intersection.Coordinate
        Constructor parameters do not match!
        Missing: [java.lang.String, java.lang.String]
        Extra:   []
```
5. Add String to double int
```
tms.intersection.Coordinate conforms
```
6. Remove default constructor, add triple int
```
tms.intersection.Coordinate conforms
```
7. Add a String argument to both single int and double int
```
tms.intersection.Coordinate does not conform:
class tms.intersection.Coordinate
    Constructor tms.intersection.Coordinate.tms.intersection.Coordinate
        Constructor parameters do not match!
        Missing: [int, java.lang.String]
        Extra:   []
```
8. Add a single String argument to single int and 2 String arguments to double int
```
tms.intersection.Coordinate does not conform:
class tms.intersection.Coordinate
    Constructor tms.intersection.Coordinate.tms.intersection.Coordinate
        Constructor parameters do not match!
        Missing: [int, java.lang.String]
        Extra:   []
```
From this line of deductive reasoning the following can be assumed as the cause:
    If (and only if) a class constructor / method (henceforth simply refered to as method) contain a default  method (that is, a method that has no arguments) AND 2 methods for which the set of arguments data types are the same, then, chalkbox will incorrectly diagnose the issue and show that one of the methods is missing and the default constructor is extra.

With some further debugging it was found that in ClassComparator in the `compareMembersOverload` method the cases where the overlaoded items are getting removed are not being handled correctly.
```
Removed from Actual Overload Left : public tms.intersection.Coordinate()
Removed from Expected overload Left : public tms.intersection.Coordinate()
Removed from Actual Overload Left : public tms.intersection.Coordinate(int)
Removed from Expected overload Left : public tms.intersection.Coordinate(int)
Removed from Actual Overload Left : public tms.intersection.Coordinate(int,int)
Removed from Expected overload Left : public tms.intersection.Coordinate(int)   # ERROR HERE, IT SHOULD MATCH
public tms.intersection.Coordinate(int,int) compared to :public tms.intersection.Coordinate()
```

For some reason, the comparator is showing no difference in the last case:
```
Removed from Actual Overload Left : public tms.intersection.Coordinate(int,int)
Removed from Expected overload Left : public tms.intersection.Coordinate(int)   # ERROR HERE, IT SHOULD MATCH
```

Doing this same test with:
6. remove default, add triple int
```
Removed from Actual Overload Left: public tms.intersection.Coordinate(int)
Removed from Expected overload Left: public tms.intersection.Coordinate(int)
Removed from Actual Overload Left: public tms.intersection.Coordinate(int,int)
Removed from Expected overload Left: public tms.intersection.Coordinate(int)
Removed from Actual Overload Left: public tms.intersection.Coordinate(int,int,int)
Removed from Expected overload Left: public tms.intersection.Coordinate(int)
public tms.intersection.Coordinate(int,int) compared to :public tms.intersection.Coordinate(int)
public tms.intersection.Coordinate(int,int,int) compared to :public tms.intersection.Coordinate(int)
```
In both cases, all elements are removed from actual Overload but in the case of 6. the comparisons are done to a constructor with an int argument while in the normal they are compared to the default.

This lead to a final test:
9. Repalce default constructor with a constructor that takes a single float.
```
tms.intersection.Coordinate does not conform:
class tms.intersection.Coordinate
    Constructor tms.intersection.Coordinate.tms.intersection.Coordinate
        Constructor parameters do not match!
        Missing: [int, int]
        Extra:   [float
```
Debugging:
```
Removed from Actual Overload Left: public tms.intersection.Coordinate(float)
Removed from Expected overload Left: public tms.intersection.Coordinate(float)
Removed from Actual Overload Left: public tms.intersection.Coordinate(int)
Removed from Expected overload Left: public tms.intersection.Coordinate(int)
Removed from Actual Overload Left: public tms.intersection.Coordinate(int,int)
Removed from Expected overload Left: public tms.intersection.Coordinate(int)
public tms.intersection.Coordinate(int,int) compared to :public tms.intersection.Coordinate(float)
```

It should hopefully now be clear that the TODO on lines 123 and 125 are useful to show that it does not in fact "work", only by coincidence most of the time.

This also highlights a potential flaw inside of the system. Inside a ListFlag the expected and actual arguments are kept track of using a single generic type `T`. In a method (and constructor) the class types are kept track of with their names.
This approach will give incorrect results IF the class of the methods are the same due to how the `isSet` method only checking if the other list CONTAINS that item and not necassarily that if it in the correct location.

Clearly here there are 2 distinct issues.

The first:
    The ListFlag `isSet` method should be more rigourous with the comparison and not just that it contains.
    In this case an approach using the NUMBER of occurences of that item was used so that the ListFlag can maintain compatability with the inclusion of different ordering of methods within a file. An approach using indexing is possible, if one cares that an index based system restricts the system to treat `Coordinate(int,String)` as different to `Coordinate(String,int)` but a seperate subclass for use by methods / constructors would be required.)

The second:
    The way that methods are being handled when it comes to overloading is incorrect as the wrong thing are being removed just due to the fact that they share common variables.
    Because this relies on how ListFlag `isSet` works the method should be fixed by the change to the first issue.

Caleb-Wishart committed:
Update ClassComparator.java

Update ClassComparator to ignore constructors.

Temporarily Fixes issue caused by overloaded constructors

Example output:
class researchsim.map.Coordinate
    Constructor researchsim.map.Coordinate.researchsim.map.Coordinate
        Constructor parameters do not match!
        Missing: [int, int]
        Extra:   []

researchsim.map.Coordinate has constructors:
Coordinate()
Coordinate(int)
Coordinate(int,int)

Caleb-Wishart committed:
Solved bug with ClassComparatar relating to Method / Constructor Overloading

Reverted last commit.

Issue with chalkbox autograder.

It appears that in situations where method / Constructor overloading occurs with only a single variable type, chalkbox will get confused with conformance.

The following tests were completed using chalkbox-0.5.3 and the following base class (placed in the packaged mentioned).

```java
package tms.intersection;

public class Coordinate {

    // default
    public Coordinate(){}
    // single int
    public Coordinate(int x){}
    // double int
    public Coordinate(int x, int y){}
    // default
    public void test(){}
    // single int
    public void test(int x){}
    // double int
    public void test(int x, int y){}
}

```
With the base class the following results were given using the JavaEngine testEngineBasic

```
tms.intersection.Coordinate does not conform:
class tms.intersection.Coordinate
    Method tms.intersection.Coordinate.test
        Method parameters do not match!
        Missing: [int, int]
        Extra:   []

    Constructor tms.intersection.Coordinate.tms.intersection.Coordinate
        Constructor parameters do not match!
        Missing: [int, int]
        Extra:   []
```

Some tests were done to try and determine the cause. (In all tests both constructors and method are affected due to the similar nature of how they are checked. A diff is left to the exercise of the reader).

1. Remove any of the 3 constructors
```
tms.intersection.Coordinate conforms
```
2. Remove single int but add triple int
```
tms.intersection.Coordinate does not conform:
class tms.intersection.Coordinate
    Constructor tms.intersection.Coordinate.tms.intersection.Coordinate
        Constructor parameters do not match!
        Missing: [int, int, int]
        Extra:   []
```
3. Change single int to single String
```
tms.intersection.Coordinate conforms
```
4. Change all ints to String
```
tms.intersection.Coordinate does not conform:
class tms.intersection.Coordinate
    Constructor tms.intersection.Coordinate.tms.intersection.Coordinate
        Constructor parameters do not match!
        Missing: [java.lang.String, java.lang.String]
        Extra:   []
```
5. Add String to double int
```
tms.intersection.Coordinate conforms
```
6. Remove default constructor, add triple int
```
tms.intersection.Coordinate conforms
```
7. Add a String argument to both single int and double int
```
tms.intersection.Coordinate does not conform:
class tms.intersection.Coordinate
    Constructor tms.intersection.Coordinate.tms.intersection.Coordinate
        Constructor parameters do not match!
        Missing: [int, java.lang.String]
        Extra:   []
```
8. Add a single String argument to single int and 2 String arguments to double int
```
tms.intersection.Coordinate does not conform:
class tms.intersection.Coordinate
    Constructor tms.intersection.Coordinate.tms.intersection.Coordinate
        Constructor parameters do not match!
        Missing: [int, java.lang.String]
        Extra:   []
```
From this line of deductive reasoning the following can be assumed as the cause:
    If (and only if) a class constructor / method (henceforth simply refered to as method) contain a default  method (that is, a method that has no arguments) AND 2 methods for which the set of arguments data types are the same, then, chalkbox will incorrectly diagnose the issue and show that one of the methods is missing and the default constructor is extra.

With some further debugging it was found that in ClassComparator in the `compareMembersOverload` method the cases where the overlaoded items are getting removed are not being handled correctly.
```
Removed from Actual Overload Left : public tms.intersection.Coordinate()
Removed from Expected overload Left : public tms.intersection.Coordinate()
Removed from Actual Overload Left : public tms.intersection.Coordinate(int)
Removed from Expected overload Left : public tms.intersection.Coordinate(int)
Removed from Actual Overload Left : public tms.intersection.Coordinate(int,int)
Removed from Expected overload Left : public tms.intersection.Coordinate(int)   # ERROR HERE, IT SHOULD MATCH
public tms.intersection.Coordinate(int,int) compared to :public tms.intersection.Coordinate()
```

For some reason, the comparator is showing no difference in the last case:
```
Removed from Actual Overload Left : public tms.intersection.Coordinate(int,int)
Removed from Expected overload Left : public tms.intersection.Coordinate(int)   # ERROR HERE, IT SHOULD MATCH
```

Doing this same test with:
6. remove default, add triple int
```
Removed from Actual Overload Left: public tms.intersection.Coordinate(int)
Removed from Expected overload Left: public tms.intersection.Coordinate(int)
Removed from Actual Overload Left: public tms.intersection.Coordinate(int,int)
Removed from Expected overload Left: public tms.intersection.Coordinate(int)
Removed from Actual Overload Left: public tms.intersection.Coordinate(int,int,int)
Removed from Expected overload Left: public tms.intersection.Coordinate(int)
public tms.intersection.Coordinate(int,int) compared to :public tms.intersection.Coordinate(int)
public tms.intersection.Coordinate(int,int,int) compared to :public tms.intersection.Coordinate(int)
```
In both cases, all elements are removed from actual Overload but in the case of 6. the comparisons are done to a constructor with an int argument while in the normal they are compared to the default.

This lead to a final test:
9. Repalce default constructor with a constructor that takes a single float.
```
tms.intersection.Coordinate does not conform:
class tms.intersection.Coordinate
    Constructor tms.intersection.Coordinate.tms.intersection.Coordinate
        Constructor parameters do not match!
        Missing: [int, int]
        Extra:   [float
```
Debugging:
```
Removed from Actual Overload Left: public tms.intersection.Coordinate(float)
Removed from Expected overload Left: public tms.intersection.Coordinate(float)
Removed from Actual Overload Left: public tms.intersection.Coordinate(int)
Removed from Expected overload Left: public tms.intersection.Coordinate(int)
Removed from Actual Overload Left: public tms.intersection.Coordinate(int,int)
Removed from Expected overload Left: public tms.intersection.Coordinate(int)
public tms.intersection.Coordinate(int,int) compared to :public tms.intersection.Coordinate(float)
```

It should hopefully now be clear that the TODO on lines 123 and 125 are useful to show that it does not in fact "work", only by coincidence most of the time.

This also highlights a potential flaw inside of the system. Inside a ListFlag the expected and actual arguments are kept track of using a single generic type `T`. In a method (and constructor) the class types are kept track of with their names.
This approach will give incorrect results IF the class of the methods are the same due to how the `isSet` method only checking if the other list CONTAINS that item and not necassarily that if it in the correct location.

Clearly here there are 2 distinct issues.

The first:
    The ListFlag `isSet` method should be more rigourous with the comparison and not just that it contains.
    In this case an approach using the NUMBER of occurences of that item was used so that the ListFlag can maintain compatability with the inclusion of different ordering of methods within a file. An approach using indexing is possible, if one cares that an index based system restricts the system to treat `Coordinate(int,String)` as different to `Coordinate(String,int)` but a seperate subclass for use by methods / constructors would be required.)

The second:
    The way that methods are being handled when it comes to overloading is incorrect as the wrong thing are being removed just due to the fact that they share common variables.
    Because this relies on how ListFlag `isSet` works the method should be fixed by the change to the first issue.

See diff for solution.

Caleb-Wishart committed:
Update to fix toString of ListFlag

Now shows the correct error messages.
The same HashSet technique could be used in the `isSet` method but the performance gain is debatable

maxcmiller commented:
Could add a unit test for this behaviour, e.g. https://github.com/maxcmiller/chalkbox/commit/977235ee5c4de1fc67f9a413b7db6b3756e83478#diff-55f662366a79f148c4f59d0d0527b9bb5ec211331b0e490b40f4f38ef39bc868

Caleb-Wishart committed:
Add test to expected

As per
https://github.com/UQTools/chalkbox/pull/8#issuecomment-1189608394

Caleb-Wishart commented:
Adds test as per above comment
Added to the expected

Caleb-Wishart committed:
Update PythonEngineTest.java

Allow for tests to be run platform independent

Caleb-Wishart commented:
Accidental close, 
Not sure why this is failing but don't know enough to figure it out.

Caleb-Wishart committed:
Update runner output to be more verbose

Fix issues of

testName: null

stacktrace

to isntead say

testName: No message given.

stacktrace

Caleb-Wishart committed:
Removed changes to line nums due to platform changes

Caleb-Wishart committed:
Add additional Coordinate test to all faulties

Caleb-Wishart committed:
Update to allow `null` as last word in message

Caleb-Wishart committed:
Update JUnitListener.java

Caleb-Wishart commented:
Made an additional change in https://github.com/UQTools/chalkbox/pull/8/commits/a9217b7278ab479b566d8502311b872ce4ce5d9a to make it such that
```
testDefaultConstructor(pizza.PizzaTest): null

java.lang.StackOverflowError
	at menu.MenuItem.registerMenuItem(MenuItem.java:11)
	at menu.MenuItem.registerMenuItem(MenuItem.java:11)
	...
```
becomes
```
testDefaultConstructor(pizza.PizzaTest): No message given, refer to stack trace.

java.lang.StackOverflowError
	at menu.MenuItem.registerMenuItem(MenuItem.java:11)
	at menu.MenuItem.registerMenuItem(MenuItem.java:11)
	...
```

Caleb-Wishart committed:
Update expected-basic.json

Caleb-Wishart commented:
@maxcmiller 
Finally fixed tests and ready for merge if you are happy! (yay)

# Key file changes
[src/chalkbox/api/common/java/JUnitListener.java](https://github.com/UQTools/chalkbox/pull/8/files#diff-de02794e904717ac8ff0903e954717f365bdd0688e84c9d361e2d7fc1727ae4e)
- Modified such that message will not be `null` --> see Failure class for reason on `equals`


[src/chalkbox/java/conformance/comparator/flags/ListFlag.java](https://github.com/UQTools/chalkbox/pull/8/files#diff-fa23ccec5f48be39b080cab9904864dab92e26a5f73fdaf18f7ea75a374ea3fb)
- Fix for ordering and contains as mentioned previously


[test/resources/csse2002/results/expected-basic.json](https://github.com/UQTools/chalkbox/pull/8/files#diff-4ed4551fca24e4732390b9baad1e59440002d40842b75d1c6bedfa3be35bf02d)
- Added additional test class, added change to message

maxcmiller commented:
Great work
Does this include the unit test I mentioned earlier?

Caleb-Wishart commented:
No, but it does get tested via the existing check against `results.json` as it is in the example submission.

Caleb-Wishart commented:
No longer required.

Already included in V2 branch which is now default.

Caleb-Wishart commented:
Cleaning up some of my stale forks / PRs.
Thanks all for the memories. Hope you are doing well
